### PR TITLE
return of the everyones favorite origin resolver

### LINKED ISSUE #653: Rework location cache
At the moment the location cache component is not much of a cache. It returns nothing if I call Get without calling the refresh before. 

We need a component to resolve public and outbound IPs as well as location information and, possibly, cache them. I'd imagine something like this:

```golang
type cache interface {
  GetCountry() string
  GetPublicIP() string // or maybe net.IP
  GetOutboundIP() string // or maybe net.IP
}
``` 

#651 introduced lookups for these for all the services independently, this component would allow for  reuse.

### PR DESCRIPTION
**Pull request by vkuznecovas**
I did remove the origin resolver but @Waldz put me on track and it's back. Also added the missing state transition publish in the connection manager.

Closes #653

### CODE REVIEW

File: core/node/node.go [hunk 1]
```diff
@@ -66,7 +66,7 @@ func (node *Node) Start() error {
 		}
 	}()
 
-	originalLocation, err := node.locationCache.DetectLocation()
+	originalLocation, err := node.locationCache.GetOrigin()
```
Comment (Waldz) [inline core/node/node.go]: What was wrong with this interface which clearly specified that original country is being set here:
```
// Cache allows caching location
type Cache interface {
	Get() Location
	RefreshAndGet() (Location, error)
}
```

Comment (vkuznecovas) [reply core/node/node.go]: Cache.Get is not discernible

File: consumer/statistics/reporter.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ type SessionStatisticsReporter struct {
 }
 
 // NewSessionStatisticsReporter function creates new session stats sender by given options
-func NewSessionStatisticsReporter(statisticsTracker StatsTracker, remoteReporter Reporter, signerFactory identity.SignerFactory, locationDetector location.Resolver, interval time.Duration) *SessionStatisticsReporter {
+func NewSessionStatisticsReporter(statisticsTracker StatsTracker, remoteReporter Reporter, signerFactory identity.SignerFactory, locationDetector location.OriginResolver, interval time.Duration) *SessionStatisticsReporter {
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply consumer/statistics/reporter.go]: We're coupled to location.Location anyway, so might as well reuse the interface. Adding an interface definition here will still need the location.Location.

File: core/connection/manager.go [hunk 3]
```diff
@@ -334,6 +334,10 @@ func (manager *connectionManager) Disconnect() error {
 	manager.cleanConnection()
 	manager.setStatus(statusNotConnected())
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply core/connection/manager.go]: I'd have to remap them there, and out of scope for this PR.

File: core/location/interface.go [hunk 4]
```diff
@@ -21,3 +21,8 @@ package location
 type Resolver interface {
 	DetectLocation() (Location, error)
 }
+
+// OriginResolver fetches the original country
+type OriginResolver interface {
+	GetOrigin() (Location, error)
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply core/location/interface.go]: We're coupled to location.Location anyway, so might as well reuse the interface. Adding an interface definition here will still need the location.Location.

File: core/location/cache.go [hunk 5]
```diff
@@ -55,11 +56,31 @@ func (c *Cache) fetchAndSave() (Location, error) {
...
+		// This won't be modified further.
+		if !c.isOriginSet() {
+			c.origin = &loc
+		}
+	}
 	return loc, err
 }
 
+func (c *Cache) isOriginSet() bool {
+	return c.origin != nil
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply core/location/cache.go]: Why do we need a lock here?

File: core/location/cache.go [hunk 6]
```diff
@@ -55,11 +56,31 @@ func (c *Cache) fetchAndSave() (Location, error) {
...
+}
+
+// GetOrigin returns the origin for the user - a location that's not modified by starting services.
+func (c *Cache) GetOrigin() (Location, error) {
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply core/location/cache.go]: I'd rather keep the methods relating to the same feature on the same component, but named differently.

**Comment by vkuznecovas** [comment]
@Waldz plix review again

**Pull request merged**